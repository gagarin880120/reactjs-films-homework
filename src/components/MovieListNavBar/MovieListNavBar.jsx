import React from 'react';
import PropTypes from 'prop-types';
import styles from './MovieListNavBar.module.scss';

export default function MovieListNavBar({
  onTrending, onTopRated, onUpcoming, genres, onGenreChange,
  currentGenre, switchViewMode, viewMode, currentURL,
}) {
  const activeButtonStyle = {
    fontWeight: 'bold',
    textDecoration: 'underline',
  };
  return (
    <div className={styles.wrapper}>
      <div className={styles.actionButtonsContainer}>
        <button
          type="button"
          id="trending"
          className={styles.button}
          style={currentURL.includes('/popular') ? activeButtonStyle : null}
          onClick={() => {
            onTrending();
          }}
        >
          Trending
        </button>
        <button
          type="button"
          id="topRated"
          style={currentURL.includes('top_rated') ? activeButtonStyle : null}
          className={styles.button}
          onClick={() => {
            onTopRated();
          }}
        >
          Top rated
        </button>
        <button
          type="button"
          id="upComing"
          className={styles.button}
          style={currentURL.includes('upcoming') ? activeButtonStyle : null}
          onClick={() => {
            onUpcoming();
          }}
        >
          Coming soon
        </button>
        <select
          className={styles.button}
          name="genres"
          style={currentURL.includes('discover') ? activeButtonStyle : null}
          onChange={(e) => {
            onGenreChange(e.target.value);
          }}
          value={currentGenre}
        >
          <option value="">Genre</option>
          {
          genres.map((genre) => (
            <option key={genre.id} value={genre.id}>{genre.name}</option>))
        }
        </select>
      </div>
      <div className={styles.viewSwitchersContainer}>
        <button
          type="button"
          id="galleryView"
          className={styles.button}
          onClick={() => {
            switchViewMode('gallery');
          }}
        >
          <i className="fas fa-th" style={viewMode === 'list' ? { color: '#D6D7D7' } : null} />
        </button>
        <button
          type="button"
          id="listView"
          className={styles.button}
          onClick={() => {
            switchViewMode('list');
          }}
        >
          <i className="fas fa-th-list" style={viewMode === 'gallery' ? { color: '#D6D7D7' } : null} />
        </button>
      </div>

    </div>
  );
}

MovieListNavBar.propTypes = {
  onTrending: PropTypes.func,
  onTopRated: PropTypes.func,
  onUpcoming: PropTypes.func,
  onGenreChange: PropTypes.func,
  genres: PropTypes.instanceOf(Array),
  currentGenre: PropTypes.string,
  switchViewMode: PropTypes.func,
  viewMode: PropTypes.string,
  currentURL: PropTypes.string,
};

MovieListNavBar.defaultProps = {
  onTrending: null,
  onTopRated: null,
  onUpcoming: null,
  onGenreChange: null,
  genres: [],
  currentGenre: '',
  switchViewMode: null,
  viewMode: '',
  currentURL: '',
};
