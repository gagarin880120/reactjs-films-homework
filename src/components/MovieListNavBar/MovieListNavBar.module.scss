.wrapper {
  display: flex;
  justify-content: space-between;

  select {
    margin-top: -5px;
  }
}

.link {
  border: none;
  background: transparent;
  cursor: pointer;
  outline: 0;
  color: #3bb3df;
  font-size: 18px;
  text-decoration: none;

  i {
    font-size: 21px;
  }
}

.linksContainer {
  width: 450px;
  display: flex;
  justify-content: space-between;
}

.viewSwitchersContainer {
  display: flex;
}

@media screen and (max-width: 600px) {
  .linksContainer {
    width: 80%;
    display: block;

    .link {
      margin: 1%;
    }
  }
}

@media screen and (max-width: 400px) {
  .linksContainer {
    .link {
      display: block;
    }
  }
}
