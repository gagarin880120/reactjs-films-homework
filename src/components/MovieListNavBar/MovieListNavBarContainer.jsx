import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MovieListNavBar from './MovieListNavBar';
import { setViewMode, setCurrentAPIRequest } from '../../redux/actions';
import {
  genresSelector, viewModeSelector, currentAPIRequestSelector,
} from '../../redux/selectors';

export function MovieListNavBarContainer({
  onTrending, onTopRated, onUpcoming, genres, onGenreChange,
  switchViewMode, viewMode, currentAPIRequest,
}) {
  return (
    <MovieListNavBar
      onTrending={onTrending}
      onTopRated={onTopRated}
      onUpcoming={onUpcoming}
      onGenreChange={onGenreChange}
      genres={genres}
      switchViewMode={switchViewMode}
      viewMode={viewMode}
      currentAPIRequest={currentAPIRequest}
    />
  );
}

export const mapStateToProps = (state) => ({
  genres: genresSelector(state),
  viewMode: viewModeSelector(state),
  currentAPIRequest: currentAPIRequestSelector(state),
});

export const mapDispatchToProps = (dispatch) => ({
  onTrending() {
    dispatch(setCurrentAPIRequest('popular'));
  },
  onTopRated() {
    dispatch(setCurrentAPIRequest('top_rated'));
  },
  onUpcoming() {
    dispatch(setCurrentAPIRequest('upcoming'));
  },
  onGenreChange(genreId) {
    dispatch(setCurrentAPIRequest(`genre=${genreId}`));
  },
  switchViewMode(mode) {
    dispatch(setViewMode(mode));
  },
});

MovieListNavBarContainer.propTypes = {
  onTrending: PropTypes.func,
  onTopRated: PropTypes.func,
  onUpcoming: PropTypes.func,
  onGenreChange: PropTypes.func,
  genres: PropTypes.instanceOf(Array),
  switchViewMode: PropTypes.func,
  viewMode: PropTypes.string,
  currentAPIRequest: PropTypes.string,
};

MovieListNavBarContainer.defaultProps = {
  onTrending: null,
  onTopRated: null,
  onUpcoming: null,
  onGenreChange: null,
  genres: [],
  switchViewMode: null,
  viewMode: '',
  currentAPIRequest: '',
};

export default connect(mapStateToProps, mapDispatchToProps)(MovieListNavBarContainer);
