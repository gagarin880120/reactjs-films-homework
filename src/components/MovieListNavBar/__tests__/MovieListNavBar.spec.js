import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import MovieListNavBar from '../MovieListNavBar';
import TestRenderer, { act } from 'react-test-renderer';

describe('MovieListNavBar component', () => {
  test('should render without crashing', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<MovieListNavBar currentURL="url/popular" viewMode="list" />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  test('should render with another styles depends on props', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<MovieListNavBar currentURL="url/upcoming" viewMode="gallery" />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  test('should render with another styles depends on props', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<MovieListNavBar currentURL="url/discover" viewMode="gallery" />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  describe('events', () => {
    const onTrending = jest.fn();
    const onTopRated = jest.fn();
    const onUpcoming = jest.fn();
    const onGenreChange = jest.fn();
    const switchViewMode = jest.fn();

    const testRenderer = TestRenderer.create(
      <MovieListNavBar
        onTrending={onTrending}
        onTopRated={onTopRated}
        onUpcoming={onUpcoming}
        onGenreChange={onGenreChange}
        switchViewMode={switchViewMode}
        currentURL="url/top_rated"
      />
    );

    describe('trending button', () => {
      test('when is clicked should call onTrending function', () => {
        const button = testRenderer.root.findByProps({id: 'trending'});

        act(() => {
          button.props.onClick();
        })

        expect(onTrending).toHaveBeenCalled();
      });
    });

    describe('topRated button', () => {
      test('when is clicked should call onTopRated function', () => {
        const button = testRenderer.root.findByProps({id: 'topRated'});

        act(() => {
          button.props.onClick();
        })

        expect(onTopRated).toHaveBeenCalled();
      });
    });

    describe('upcoming button', () => {
      test('when is clicked should call onUpcoming function', () => {
        const button = testRenderer.root.findByProps({id: 'upComing'});

        act(() => {
          button.props.onClick();
        })

        expect(onUpcoming).toHaveBeenCalled();
      });
    });

    describe('select', () => {
      test('when is onChange should call onGenreChange function', () => {
        const select = testRenderer.root.findByType('select');

        act(() => {
          select.props.onChange({target: { value: '28' }});
        })

        expect(onGenreChange).toHaveBeenCalled();
      });
    });

    describe('galleryView button', () => {
      test('when is clicked should call switchViewMode function', () => {
        const button = testRenderer.root.findByProps({id: 'galleryView'});

        act(() => {
          button.props.onClick();
        })

        expect(switchViewMode).toHaveBeenCalled();
      });
    });

    describe('listView button', () => {
      test('when is clicked should call switchViewMode function', () => {
        const button = testRenderer.root.findByProps({id: 'listView'});

        act(() => {
          button.props.onClick();
        })

        expect(switchViewMode).toHaveBeenCalled();
      });
    });

  });
});
