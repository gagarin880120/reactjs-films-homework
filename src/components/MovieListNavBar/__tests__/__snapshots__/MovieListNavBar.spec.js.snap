// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieListNavBar component should render with another styles depends on props 1`] = `
<div
  className="wrapper"
>
  <div
    className="actionButtonsContainer"
  >
    <button
      className="button"
      id="trending"
      onClick={[Function]}
      style={null}
      type="button"
    >
      Trending
    </button>
    <button
      className="button"
      id="topRated"
      onClick={[Function]}
      style={null}
      type="button"
    >
      Top rated
    </button>
    <button
      className="button"
      id="upComing"
      onClick={[Function]}
      style={
        Object {
          "fontWeight": "bold",
          "textDecoration": "underline",
        }
      }
      type="button"
    >
      Coming soon
    </button>
    <select
      className="button"
      name="genres"
      onChange={[Function]}
      style={null}
      value=""
    >
      <option
        value=""
      >
        Genre
      </option>
    </select>
  </div>
  <div
    className="viewSwitchersContainer"
  >
    <button
      className="button"
      id="galleryView"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-th"
        style={null}
      />
    </button>
    <button
      className="button"
      id="listView"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-th-list"
        style={
          Object {
            "color": "#D6D7D7",
          }
        }
      />
    </button>
  </div>
</div>
`;

exports[`MovieListNavBar component should render with another styles depends on props 2`] = `
<div
  className="wrapper"
>
  <div
    className="actionButtonsContainer"
  >
    <button
      className="button"
      id="trending"
      onClick={[Function]}
      style={null}
      type="button"
    >
      Trending
    </button>
    <button
      className="button"
      id="topRated"
      onClick={[Function]}
      style={null}
      type="button"
    >
      Top rated
    </button>
    <button
      className="button"
      id="upComing"
      onClick={[Function]}
      style={null}
      type="button"
    >
      Coming soon
    </button>
    <select
      className="button"
      name="genres"
      onChange={[Function]}
      style={
        Object {
          "fontWeight": "bold",
          "textDecoration": "underline",
        }
      }
      value=""
    >
      <option
        value=""
      >
        Genre
      </option>
    </select>
  </div>
  <div
    className="viewSwitchersContainer"
  >
    <button
      className="button"
      id="galleryView"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-th"
        style={null}
      />
    </button>
    <button
      className="button"
      id="listView"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-th-list"
        style={
          Object {
            "color": "#D6D7D7",
          }
        }
      />
    </button>
  </div>
</div>
`;

exports[`MovieListNavBar component should render without crashing 1`] = `
<div
  className="wrapper"
>
  <div
    className="actionButtonsContainer"
  >
    <button
      className="button"
      id="trending"
      onClick={[Function]}
      style={
        Object {
          "fontWeight": "bold",
          "textDecoration": "underline",
        }
      }
      type="button"
    >
      Trending
    </button>
    <button
      className="button"
      id="topRated"
      onClick={[Function]}
      style={null}
      type="button"
    >
      Top rated
    </button>
    <button
      className="button"
      id="upComing"
      onClick={[Function]}
      style={null}
      type="button"
    >
      Coming soon
    </button>
    <select
      className="button"
      name="genres"
      onChange={[Function]}
      style={null}
      value=""
    >
      <option
        value=""
      >
        Genre
      </option>
    </select>
  </div>
  <div
    className="viewSwitchersContainer"
  >
    <button
      className="button"
      id="galleryView"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-th"
        style={
          Object {
            "color": "#D6D7D7",
          }
        }
      />
    </button>
    <button
      className="button"
      id="listView"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-th-list"
        style={null}
      />
    </button>
  </div>
</div>
`;
